diff --git a/subprojects/gst-plugins-bad/gst-libs/gst/webrtc/nice/nice.c b/subprojects/gst-plugins-bad/gst-libs/gst/webrtc/nice/nice.c
index d93d742eeb..d1e2b55735 100644
--- a/subprojects/gst-plugins-bad/gst-libs/gst/webrtc/nice/nice.c
+++ b/subprojects/gst-plugins-bad/gst-libs/gst/webrtc/nice/nice.c
@@ -1680,6 +1680,7 @@ gst_webrtc_nice_constructed (GObject * object)
 
   options |= NICE_AGENT_OPTION_ICE_TRICKLE;
   options |= NICE_AGENT_OPTION_REGULAR_NOMINATION;
+  options |= NICE_AGENT_OPTION_SUPPORT_RENOMINATION;
 
 /*  https://gitlab.freedesktop.org/libnice/libnice/-/merge_requests/257 */
 #if HAVE_LIBNICE_CONSENT_FIX
